### PR TITLE
Create feed impression and click events

### LINKED ISSUE #19924: [1] Create feed event and clicks in feed
## Is this feature related to a problem? Please describe.

As a user, I want my feed to represent content I'm most likely to enjoy and engage with.

As a platform, we track net popularity, but don't track feed success rate, leading to having a harder time matching people with relevant content to them.

## Describe the solution you’d like

Begin tracking feed events for articles, similarly to how we track billboard events, ultimately allowing us to tabulate a success score.

However, we'll track success more deeply than just click-through-rate. We'll also map positive reactions and comments to feed events, to track net engagement per impression for an article's feed appearance.

We will begin by simply tabulating a success score which can be included as a factor in feed relevancy, but in the future could be very useful data in a potential evolution of the algorithm.

### Specifications

A **feed event** of category "impression" should be logged for every article card that scrolls into view in the home feed, similar to how we track billboard impressions. One impression event should be logged for every article that comes into view, but if possible these should be sent in batches to avoid excess pings (i.e., batch impressions of multiple articles scrolled into view together)

A **click** should be tracked when a user clicks in from the feed.

We should map each event to the current user — we do not need to track these events for anonymous sessions. This will only be for signed in [users:](url)

We should also record which position in the feed each event occurs — ignoring interstitials like billboards (starting with position 1, not 0 😄). We will **not** account for podcasts when determining the position in the feed.

## Definition of Done

- [ ] Create new model
- [ ] Create controller and client side tracking for impressions and clicks
- [ ] Add a table to store the events
- [ ] Add endpoints for the impressions endpoint
- [ ] Batch impressions in the JS before calling the endpoint

This is part of a series of issues, this being the first one. Others that follow:
2. [[2] Track reactions from article feed clicks](https://github.com/forem/forem/issues/19925)
3. [[3] Track comments from article feed click](https://github.com/forem/forem/issues/19927)
4. [[4] Create `feed_success_score` summation attribute on article](https://github.com/forem/forem/issues/19926)

 
## Additional Context
This is inspired by the billboard events functionality so we should build on the patterns that work, but this is paired down in some ways (only tracking signed in, for example), and extended in others (tracking reaction and comment events)

This is [currently how we implement the billboard impressions and clicks through JavaScript](https://github.com/forem/forem/blob/Ridhwana/tag-page-buttons/app/javascript/packs/billboardAfterRenderActions.js) and as per the description we may want to do something similar but we'll need to do some more extensive logic for clicks where we tracked where the click is coming from and how recent it was.

We'll need to add a table for the events and add endpoints for the impressions endpoint. I'm assuming that we'll need to store data like:

user_id, article_id, article_position, category, context type

## Additional Context

- Original issue https://github.com/forem/forem/issues/19773

## Team

Please @benhalpern for any questions on this issue.

### PR DESCRIPTION
**Pull request by filleduchaos**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This one is a bit large, but lays a bit of a foundation for follow-up reaction and comment events. I have also split off bumping event counters and periodically compacting feed events to a different branch

## Related Tickets & Documents

- Related Issue #19924
- Closes #19924 

## QA Instructions, Screenshots, Recordings

https://www.loom.com/share/b4847236745b496ba3bf3c44726b8076

I did not mention in the video that if you scroll back and forth you should see that new events don't get sent from the same feed/tab. Also, if you open a few tabs and scroll within them at the same time (logged in to the same account) you should see that only one impression event per article gets recorded (as documented this is time-windowed, and will only prevent duplicates within the previous five minutes to reduce spam and noise)

### UI accessibility checklist

Made UI changes but they are background behaviour changes only.

## Added/updated tests?

- [x] Yes

## [optional] Are there any post deployment tasks we need to perform?

Nope

## [optional] What gif best describes this PR or how it makes you feel?

tired-exhausted

### CODE REVIEW

File: app/javascript/packs/homePageFeed.jsx [hunk 1]
```diff
@@ -150,7 +150,7 @@ export const renderFeed = async (timeFrame, afterRender) => {
     }
 
     return (
-      <div>
+      <Fragment>
```
Comment (filleduchaos) [inline app/javascript/packs/homePageFeed.jsx:153]: Switched to a Fragment here to eliminate the filler div and make querying the children articles more straightforward

**Review by benhalpern — APPROVED** [review]
Looks great to me, and works great.

Non-blocking optimization to consider in follow-ups: I think `feedEvents.js` could be a conditional dynamic import which only needs to trigger for signed-in contexts.

File: cypress/support/commands.js [hunk 2]
```diff
@@ -26,6 +26,26 @@ import { getInterceptsForLingeringUserRequests } from '../util/networkUtils';
...
+   */
+  (commands) =>
+    commands.reduce(
+      (allResults, command) =>
+        allResults.then((results) =>
+          command.then((result) => [...results, result]),
+        ),
+      cy.wrap([]),
+    ),
+);
```
Comment (maestromac) [inline cypress/support/commands.js:47]: Nice addition here!

**Review by maestromac — APPROVED** [review]
LGTM! Great implementation! I'm giving a high level approval to unblock and may follow up with some comments.

**Pull request merged**